### PR TITLE
YCSB 0.7.0 with cassandra2-cql binding, Cassandra deployed using binaries

### PR DESCRIPTION
**Pull request by kamsz**
- Added support for YCSB 0.7.0 with cassandra2-cql binding #872 #865 
- Instead of compiling Cassandra, using precompiled binaries from Apache
- Modify /etc/hosts to include an entry with internal ip address of virtual machine.

### CODE REVIEW

**Comment by voellm** [comment]
Thank for the contribution!  Once you have the CLA signed we can review the code and complete the pull  request.  The CLA basically says the code you are contributing is yours to give.

@yuyantingzero - Looks like the move to 0.7.0 happened!  Can you handle this code review?  Also once it is reviewed and ready kick off the integration tests with "test this please" as the sole comment.

**Comment by kamsz** [comment]
I signed it!

File: perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py [hunk 1]
```diff
@@ -50,10 +50,9 @@
 
 # TODO: Add flags.
 REPLICATION_FACTOR = 3
+COLUMN_FAMILY = 'data'
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py]: COLUMN_FAMILY constant wasn't used.

File: perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2 [hunk 2]
```diff
@@ -12,10 +12,17 @@
...
-create keyspace {{ keyspace }}
-  with placement_strategy = 'SimpleStrategy'
-  and strategy_options = {replication_factor: {{ replication_factor }} };
-
-use {{ keyspace }};
-
-create column family {{ column_family }};
+create keyspace ycsb WITH REPLICATION = {'class' : 'SimpleStrategy', 'replication_factor': {{replication_factor}} };
+USE ycsb;
+create table usertable (
```
Comment (yuyantingzero) [inline perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2]: Seems like table schema is hard-coded using this script, while ycsb package should be able to create tables with different number of fields: https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/e75b755b70b93a96e507a410447e7a4e59ff4dfe/perfkitbenchmarker/linux_packages/ycsb.py#L476

**Comment by yuyantingzero** [comment]
Since this PR updates both cassandra and ycsb. I will run a few tests after the code review, which might take some time.
cassandra-ycsb
hbase-ycsb
bigtable-ycsb
aerospike-ycsb
cloud-bigtable-ycsb
mongodb-ycsb
redis-ycsb

File: perfkitbenchmarker/linux_packages/cassandra.py [hunk 3]
```diff
@@ -103,6 +104,8 @@ def Configure(vm, seed_vms):
              'num_cpus': vm.num_cpus,
              'cluster_name': 'Test cluster'}
 
+  vm.UpdateHosts()
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_packages/cassandra.py]: Is this required to use cqlsh?

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 4]
```diff
@@ -187,6 +187,15 @@ def SetupProxy(self):
     if commands:
       self.RemoteCommand(";".join(commands))
 
+  def UpdateHosts(self):
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_virtual_machine.py]: Please rename the method to be more descriptive or make the method more generic. e.g. UpdateHostsFille(self, ip, hostname)

File: perfkitbenchmarker/linux_packages/cassandra.py (hunk 3 above)
Comment (kamsz) [inline perfkitbenchmarker/linux_packages/cassandra.py]: Cassandra refuses to start without proper /etc/hosts entry

File: perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py (hunk 1 above)
Comment (kamsz) [inline perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py]: It's used in line 165 and passed to YCSB execution params:
https://github.com/kamsz/PerfKitBenchmarker/blob/cassandra_ycsb/perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py#L165

File: perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2 (hunk 2 above)
Comment (kamsz) [inline perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2]: What do you mean by that? I should add a param to execution like --ycsb_field_count that will generate fields in table schema (with 10 as default)?

File: perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py (hunk 1 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py]: You are right, I missed line 165.

File: perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2 (hunk 2 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2]: YCSB itself will load tables with field count according to workload file ( https://github.com/brianfrankcooper/YCSB/blob/master/workloads/workload_template#L53) and we can also inject YCSB parameters to override YCSB workload file with ycsb_load_parameters and ycsb_run_parameters.

File: perfkitbenchmarker/linux_virtual_machine.py [hunk 5]
```diff
@@ -187,6 +187,15 @@ def SetupProxy(self):
     if commands:
       self.RemoteCommand(";".join(commands))
 
+  def UpdateVMHostsFile(self):
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_virtual_machine.py]: Thanks for the renaming!

File: perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py (hunk 1 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py]: But from my understanding "column family" means "table" in Cassandra. Which means the code in 165 actually created a table named "data", but the new create-ycsb-table.cql created a table named "usertable"

**Comment by stfeng** [comment]
FYI - The cut-off date for next release is EOD Thursday 3/31.

File: perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2 (hunk 2 above)
Comment (kamsz) [inline perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2]: schema is generated dynamically now

**Comment by kamsz** [comment]
I've fixed schema generation.

File: perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2 (hunk 2 above)
Comment (yuyantingzero) [inline perfkitbenchmarker/data/cassandra/create-ycsb-table.cql.j2]: Thanks for updating!

File: perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py [hunk 6]
```diff
@@ -122,6 +127,7 @@ def Prepare(benchmark_spec):
     benchmark_spec: The benchmark specification. Contains all data that is
         required to run the benchmark.
   """
+  print benchmark_spec
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_benchmarks/cassandra_ycsb_benchmark.py]: Please remove this line.

File: perfkitbenchmarker/linux_packages/cassandra.py [hunk 7]
```diff
@@ -103,6 +104,8 @@ def Configure(vm, seed_vms):
              'num_cpus': vm.num_cpus,
              'cluster_name': 'Test cluster'}
 
+  vm.UpdateVMHostsFile()
```
Comment (yuyantingzero) [inline perfkitbenchmarker/linux_packages/cassandra.py]: I cloned your branch into my local machine. 
I am able to start Cassandra cluster without updating host file. 
What is the specific error message you see? You should be able to find logs in /tmp/pkb/apache-cassandra-2.1.13/logs/system.log

**Comment by yuyantingzero** [comment]
Looks good in general. I will kick off a few side-by-side runs.

File: perfkitbenchmarker/linux_packages/cassandra.py (hunk 7 above)
Comment (kamsz) [inline perfkitbenchmarker/linux_packages/cassandra.py]: As far as I remember I've been receiving
```
Error: Exception thrown by the agent : java.net.MalformedURLException: Local host name unknown: java.net.UnknownHostException
```
I'm not able to test it at the moment, I'll try ASAP.

**Comment by kamsz** [comment]
@yuyantingzero any progress?

**Comment by yuyantingzero** [comment]
@kamsz runs finished. Looks good.
Could you sync up your branch? Also how about the vm host file issue? Have you made any progress?

**Comment by yuyantingzero** [comment]
@kamsz , any updates on this pull request, ready to check in :)?
YCSB version in PerfkitBenchmarker is 0.3.0, which is a bit outdated. If you are getting too busy, I can take over this PR and bump YCSB version to 0.9.0 by end of this month, if you do not mind.

**Comment by kamsz** [comment]
@yuyantingzero go ahead, I don't mind :)

**Comment by tedsta** [comment]
@yuyantingzero hey I see [ycsb did get updated to 0.9.0](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/a78d7d520407248c8562c54ccaaf8301819ee7ca/perfkitbenchmarker/linux_packages/ycsb.py#L61). Is this ok to close?

**Comment by yuyantingzero** [comment]
@tedsta seems good to me.

**Pull request closed without merging**